File: backend/filings/edgar.py
"""SEC EDGAR client.

EDGAR rate limits: 10 req/sec, mandatory User-Agent. We batch fetches with a
hand-rolled token-bucket throttle and cache aggressively (10-Ks are annual).
"""
import json
import time
import threading
from dataclasses import dataclass
from pathlib import Path
import urllib.request
import urllib.error


SEC_USER_AGENT = "AlphaDesk research@alphadesk.app"
EDGAR_BASE = "https://www.sec.gov"
EDGAR_DATA = "https://data.sec.gov"

_CACHE_DIR = Path(__file__).resolve().parent.parent / "data" / "filings"
_CIK_FILE = _CACHE_DIR / "_company_tickers.json"
_CIK_TTL_SECONDS = 30 * 24 * 3600  # refresh monthly


@dataclass
class FilingMeta:
    cik: str             # zero-padded 10-digit
    accession: str       # e.g. 0000320193-24-000123
    primary_doc: str     # filename of the primary 10-K document
    filing_date: str     # YYYY-MM-DD
    fiscal_year_end: str # YYYY-MM-DD
    url: str             # full URL to the primary document


# ── Rate limiting ──────────────────────────────────────────────────────────
_lock = threading.Lock()
_last_call = 0.0
_MIN_INTERVAL = 0.11  # ~9 req/sec, comfortably under SEC's 10/sec limit


def _throttle():
    global _last_call
    with _lock:
        now = time.time()
        wait = _MIN_INTERVAL - (now - _last_call)
        if wait > 0:
            time.sleep(wait)
        _last_call = time.time()


def _http_get(url: str, timeout: int = 20) -> bytes:
    _throttle()
    req = urllib.request.Request(url, headers={
        "User-Agent": SEC_USER_AGENT,
        "Accept-Encoding": "gzip, deflate",
        "Host": _host_for(url),
    })
    with urllib.request.urlopen(req, timeout=timeout) as resp:
        data = resp.read()
        if resp.headers.get("Content-Encoding") == "gzip":
            import gzip
            data = gzip.decompress(data)
        elif resp.headers.get("Content-Encoding") == "deflate":
            import zlib
            data = zlib.decompress(data)
        return data


def _host_for(url: str) -> str:
    if url.startswith(EDGAR_DATA):
        return "data.sec.gov"
    return "www.sec.gov"


# ── CIK lookup ─────────────────────────────────────────────────────────────

def _load_cik_table() -> dict:
    """Return ticker -> 10-digit CIK mapping. Cached locally and refreshed monthly."""
    _CACHE_DIR.mkdir(parents=True, exist_ok=True)
    if _CIK_FILE.exists() and (time.time() - _CIK_FILE.stat().st_mtime) < _CIK_TTL_SECONDS:
        raw = json.loads(_CIK_FILE.read_text())
    else:
        data = _http_get(f"{EDGAR_BASE}/files/company_tickers.json")
        raw = json.loads(data.decode())
        _CIK_FILE.write_text(json.dumps(raw))

    table = {}
    for entry in raw.values():
        cik = str(entry["cik_str"]).zfill(10)
        table[entry["ticker"].upper()] = cik
    return table


def cik_for_ticker(ticker: str) -> str | None:
    return _load_cik_table().get(ticker.upper())


# ── Filing index ───────────────────────────────────────────────────────────

def _submissions(cik: str) -> dict:
    cache = _CACHE_DIR / cik / "submissions.json"
    if cache.exists() and (time.time() - cache.stat().st_mtime) < 24 * 3600:
        return json.loads(cache.read_text())
    data = _http_get(f"{EDGAR_DATA}/submissions/CIK{cik}.json")
    cache.parent.mkdir(parents=True, exist_ok=True)
    cache.write_text(data.decode())
    return json.loads(data.decode())


def fetch_two_latest_10k(ticker: str) -> list[FilingMeta]:
    """Return the two most recent 10-K filings for `ticker` (newest first).

    Returns [] if the ticker is unknown or has fewer than one 10-K on file.
    """
    cik = cik_for_ticker(ticker)
    if not cik:
        return []
    submissions = _submissions(cik)
    recent = submissions.get("filings", {}).get("recent", {})
    if not recent:
        return []

    forms = recent.get("form", [])
    accessions = recent.get("accessionNumber", [])
    primaries = recent.get("primaryDocument", [])
    filing_dates = recent.get("filingDate", [])
    period_dates = recent.get("reportDate", [])

    out: list[FilingMeta] = []
    for i, form in enumerate(forms):
        if form != "10-K":
            continue
        acc = accessions[i]
        acc_no_dash = acc.replace("-", "")
        primary = primaries[i]
        url = f"{EDGAR_BASE}/Archives/edgar/data/{int(cik)}/{acc_no_dash}/{primary}"
        out.append(FilingMeta(
            cik=cik,
            accession=acc,
            primary_doc=primary,
            filing_date=filing_dates[i],
            fiscal_year_end=period_dates[i] if i < len(period_dates) else "",
            url=url,
        ))
        if len(out) >= 2:
            break
    return out


# ── Document fetch + cache ─────────────────────────────────────────────────

def fetch_filing_html(meta: FilingMeta) -> str:
    """Download and cache the raw HTML of a 10-K filing."""
    cache = _CACHE_DIR / meta.cik / meta.accession.replace("-", "") / "primary.html"
    if cache.exists():
        return cache.read_text(encoding="utf-8", errors="replace")
    data = _http_get(meta.url)
    text = data.decode("utf-8", errors="replace")
    cache.parent.mkdir(parents=True, exist_ok=True)
    cache.write_text(text, encoding="utf-8")
    return text

File: netlify/functions/shared/edgar.mjs
/**
 * Shared SEC EDGAR client for Netlify functions (filings, insider, etc.).
 *
 * Ported from backend/filings/edgar.py. Same throttle (~9 req/sec), same
 * required User-Agent, same logical caches — but the disk-backed cache is
 * replaced with Netlify Blobs because functions are stateless across cold
 * starts.
 *
 * What's cached in Blobs:
 *   - CIK lookup table (~500KB, refreshed every 30 days). Avoids a 500KB
 *     download on every invocation.
 *   - Per-CIK submissions index (24h TTL). Avoids hammering data.sec.gov
 *     when the same ticker is hit multiple times in quick succession.
 *
 * What's NOT cached: filing HTMLs. They're large and only fetched 2x per
 * filings analysis. Each caller (filings.mjs, insider.mjs) caches its own
 * scored output instead.
 */
import { getStore } from '@netlify/blobs';

export const SEC_USER_AGENT = 'AlphaDesk research@alphadesk.app';
export const EDGAR_BASE = 'https://www.sec.gov';
export const EDGAR_DATA = 'https://data.sec.gov';

const CIK_BLOB_STORE = 'edgar-cache';
const CIK_BLOB_KEY   = 'cik_table';
const CIK_TTL_MS     = 30 * 24 * 60 * 60 * 1000;

const SUBMISSIONS_BLOB_STORE = 'edgar-cache';
const SUBMISSIONS_TTL_MS     = 24 * 60 * 60 * 1000;

// ── Throttle ───────────────────────────────────────────────────────────────
// Per-invocation only (function instances are short-lived). Multiple
// concurrent invocations could collectively exceed 10/sec, but each
// individual function makes few EDGAR calls so the steady-state load is low.
const MIN_INTERVAL_MS = 110;  // ~9 req/sec, comfortably under SEC's 10/sec cap
let lastCallAt = 0;

async function throttle() {
  const now = Date.now();
  const wait = MIN_INTERVAL_MS - (now - lastCallAt);
  if (wait > 0) await new Promise(r => setTimeout(r, wait));
  lastCallAt = Date.now();
}

function hostFor(url) {
  return url.startsWith(EDGAR_DATA) ? 'data.sec.gov' : 'www.sec.gov';
}

/**
 * Throttled EDGAR HTTPS fetch. Returns the Response — caller decides whether
 * to read as JSON, text, or buffer. Throws on non-2xx so callers can decide
 * whether to swallow or surface the error.
 */
export async function edgarGet(url, { timeoutMs = 15000 } = {}) {
  await throttle();
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);
  try {
    const res = await fetch(url, {
      signal: controller.signal,
      headers: {
        'User-Agent':    SEC_USER_AGENT,
        'Accept':        '*/*',
        'Host':          hostFor(url),
      },
    });
    if (!res.ok) {
      throw new Error(`EDGAR ${res.status} ${res.statusText}: ${url}`);
    }
    return res;
  } finally {
    clearTimeout(timer);
  }
}

// ── CIK lookup ─────────────────────────────────────────────────────────────

/**
 * Returns { TICKER: '0000320193', ... } for all SEC-registered tickers.
 * The raw company_tickers.json is ~500KB; we transform once and cache the
 * compact ticker→cik mapping.
 */
export async function getCikTable() {
  let store;
  try { store = getStore(CIK_BLOB_STORE); } catch { store = null; }

  if (store) {
    try {
      const cached = await store.get(CIK_BLOB_KEY, { type: 'json' });
      if (cached?.cached_at &&
          Date.now() - new Date(cached.cached_at).getTime() < CIK_TTL_MS) {
        return cached.table;
      }
    } catch { /* cache miss → fall through */ }
  }

  const res = await edgarGet(`${EDGAR_BASE}/files/company_tickers.json`);
  const raw = await res.json();

  const table = {};
  for (const entry of Object.values(raw)) {
    if (entry?.ticker && entry.cik_str !== undefined) {
      table[String(entry.ticker).toUpperCase()] =
        String(entry.cik_str).padStart(10, '0');
    }
  }

  if (store) {
    try {
      await store.set(CIK_BLOB_KEY, JSON.stringify({
        table,
        cached_at: new Date().toISOString(),
      }));
    } catch { /* silent */ }
  }
  return table;
}

export async function cikForTicker(ticker) {
  const table = await getCikTable();
  return table[String(ticker).toUpperCase()] || null;
}

// ── Submissions index ─────────────────────────────────────────────────────

/**
 * Fetch /submissions/CIK{cik}.json. Cached per-CIK with 24h TTL.
 * Returns the full submissions object — caller filters by form type.
 */
export async function fetchSubmissions(cik) {
  let store;
  try { store = getStore(SUBMISSIONS_BLOB_STORE); } catch { store = null; }
  const cacheKey = `submissions_${cik}`;

  if (store) {
    try {
      const cached = await store.get(cacheKey, { type: 'json' });
      if (cached?.cached_at &&
          Date.now() - new Date(cached.cached_at).getTime() < SUBMISSIONS_TTL_MS) {
        return cached.data;
      }
    } catch { /* miss */ }
  }

  const res = await edgarGet(`${EDGAR_DATA}/submissions/CIK${cik}.json`);
  const data = await res.json();

  if (store) {
    try {
      await store.set(cacheKey, JSON.stringify({
        data,
        cached_at: new Date().toISOString(),
      }));
    } catch { /* silent */ }
  }
  return data;
}

// ── Filing fetch ──────────────────────────────────────────────────────────

/**
 * @typedef {Object} FilingMeta
 * @property {string} cik             zero-padded 10-digit
 * @property {string} accession       e.g. 0000320193-24-000123
 * @property {string} primary_doc     filename of the primary document
 * @property {string} filing_date     YYYY-MM-DD
 * @property {string} fiscal_year_end YYYY-MM-DD or ''
 * @property {string} url             full URL to the primary document
 */

export function buildFilingUrl(cik, accession, primaryDoc) {
  const accNoDash = accession.replace(/-/g, '');
  return `${EDGAR_BASE}/Archives/edgar/data/${parseInt(cik, 10)}/${accNoDash}/${primaryDoc}`;
}

/**
 * Walk a submissions payload and return up to N most-recent filings matching
 * `formType` (e.g. "10-K", "4"). Newest first because EDGAR's recent[] array
 * is already sorted newest-first.
 */
export function pickRecentFilings(submissions, formType, max = 2, opts = {}) {
  const { sinceDate = null } = opts;
  const recent = submissions?.filings?.recent || {};
  const forms = recent.form || [];
  const accs = recent.accessionNumber || [];
  const primaries = recent.primaryDocument || [];
  const dates = recent.filingDate || [];
  const periods = recent.reportDate || [];

  const out = [];
  const cik = submissions?.cik
    ? String(submissions.cik).padStart(10, '0')
    : null;

  for (let i = 0; i < forms.length; i++) {
    if (forms[i] !== formType) continue;
    if (sinceDate && dates[i] && dates[i] < sinceDate) break;

    const acc = accs[i];
    const primary = primaries[i];
    if (!acc || !primary || !cik) continue;

    out.push({
      cik,
      accession:        acc,
      primary_doc:      primary,
      filing_date:      dates[i] || '',
      fiscal_year_end:  periods[i] || '',
      url:              buildFilingUrl(cik, acc, primary),
    });
    if (out.length >= max) break;
  }
  return out;
}

/**
 * Fetch a single filing's primary document body. Returns raw HTML/XML text.
 * Not cached at this layer — filings.mjs and insider.mjs cache their scored
 * results instead.
 */
export async function fetchFilingText(meta) {
  const res = await edgarGet(meta.url, { timeoutMs: 25000 });
  return res.text();
}
